non-critical-infra/modules/backup.nix

```diff
@@ -0,0 +1,145 @@
+      description = ''
+        Hostname of the SSH remote host.
+      '';
+    };
+
+    hostPublicKey = mkOption {
+      type = str;
+      example = "ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEA5EB5p/5Hp3hGW1oHok+PIOH9Pbn7cnUiGmUEBrCVjnAw+HrKyN8bYVV0dIGllswYXwkG/+bgiBlE6IVIBAq+JwVWu1Sss3KarHY3OvFJUXZoZyRRg/Gc/+LRCE7lyKpwWQ70dbelGRyyJFH36eNv6ySXoUYtGkwlU5IVaHPApOxe4LHPZa/qhSRbPo2hwoh0orCtgejRebNtW5nlx00DNFgsvn8Svz2cIYLxsPVzKgUxs8Zxsxgn+Q/UvR7uq4AbAhyBMLxv7DjJ1pc7PJocuTno2Rw9uMZi1gkjbnmiOh6TTXIEWbnroyIhwc8555uto9melEUmWNQ+C+PwAK+MPw==";
+      description = ''
+        Public SSH host key of the remote host. Discoverable using e.g. `ssh-keyscan`.
+      '';
+    };
+
+    port = mkOption {
+      type = port;
+      default = 22;
+      description = ''
+        Port of the SSH remote host.
+      '';
+      apply = toString;
+    };
+
+    quota = mkOption {
+      type = nullOr str;
+      default = null;
+      example = "90G";
+      description = ''
+        Quota for the borg repository. Useful to prevent the target disk from running full and ensuring borg keeps some space to work with.
+      '';
+    };
+
+    includes = mkOption {
+      type = listOf path;
+      default = [];
+      description = ''
+        Paths to include in the backup.
+      '';
+    };
+
+    excludes = mkOption {
+      type = listOf path;
+      default = [];
+      description = ''
+        Paths to exclude in the backup.
+      '';
+    };
+
+    preHook = mkOption {
+      type = lines;
+      default = "";
+      description = ''
+        Shell commands to run before the backup.
+      '';
+    };
+
+    postHook = mkOption {
+      type = lines;
+      default = "";
+      description = ''
+        Shell commands to run after the backup.
+      '';
+    };
+
+    wantedUnits = mkOption {
+      type = listOf str;
+      default = [];
+      description = ''
+        List of units to require before starting the backup.
+      '';
+    };
+  };
+
+  config = lib.mkIf (cfg.includes != []) {
+    programs.ssh.knownHosts."${cfg.host}" = {
+      publicKey = "${cfg.hostPublicKey}";
+    };
+
+    systemd.services.borgbackup-job-state = {
+      wants = cfg.wantedUnits;
+      after = cfg.wantedUnits;
+    };
+
+    systemd.timers.borgbackup-job-state.timerConfig = {
+      # Spread all backups over the day
+      RandomizedDelaySec = "24h";
+      FixedRandomDelay = true;
+    };
+
+    services.borgbackup.jobs.state = {
+      inherit (cfg) preHook postHook;
+
+      # Create the repo     
+      doInit = true;
+
+      # Create daily backups, but prune to a reasonable amount
+      startAt = "daily";
+      prune.keep = {
+        daily = 7;
+        weekly = 4;
+        monthly = 3;
+      };
+
+      # What to backup
+      paths = cfg.includes;
+      exclude = cfg.excludes;
+
+      # Where to backup it to
+      repo = "${cfg.user}@${cfg.host}:${config.networking.fqdn}";
+      environment.BORG_RSH = "ssh -p ${cfg.port} -i ~/.ssh/id_ed25519";
```

Comment:
I think we can use the local root ssh key and not manage it any further. The ssh key install also happened manually.

https://docs.hetzner.com/robot/storage-box/backup-space-ssh-keys

Reply:
Can we have this use an SSH private key provisioned by sops instead?

Reply:
Done in #345.